File: validation/cases/analysis_poles/analysis_poles.m
% analysis_poles.m -- Reference system analysis via Octave Control Toolbox.
% Poles, stability, controllability, observability for continuous and discrete systems.
%
% Usage: octave --no-gui analysis_poles.m > analysis_poles_octave.csv

pkg load control;

% Continuous mass-spring-damper
A_c = [0, 1; -1, -0.5];
B_c = [0; 1];
C_c = [1, 0];

sys_c = ss(A_c, B_c, C_c, 0);
p_c = pole(sys_c);

% Discrete version
sys_d = c2d(sys_c, 0.05);
p_d = pole(sys_d);

% Controllability rank
ctrb_rank = rank(ctrb(sys_d.A, sys_d.B));
obsv_rank = rank(obsv(sys_d.A, sys_d.C));

% Stability checks: 1.0 for stable, 0.0 for unstable
is_stable_c = all(real(p_c) < 0);
is_stable_d = all(abs(p_d) < 1);

fprintf('pc0_re,pc0_im,pc1_re,pc1_im,pd0_re,pd0_im,pd1_re,pd1_im,ctrb_rank,obsv_rank,stable_c,stable_d\n');

% Sort poles by imaginary part (positive first)
[~, idx_c] = sort(imag(p_c), 'descend');
p_c = p_c(idx_c);
[~, idx_d] = sort(imag(p_d), 'descend');
p_d = p_d(idx_d);

fprintf('%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e\n', ...
        real(p_c(1)), imag(p_c(1)), real(p_c(2)), imag(p_c(2)), ...
        real(p_d(1)), imag(p_d(1)), real(p_d(2)), imag(p_d(2)), ...
        ctrb_rank, obsv_rank, is_stable_c, is_stable_d);

File: validation/cases/analysis_poles/analysis_poles.cpp
// analysis_poles.cpp -- System analysis matching analysis_poles.m
// Usage: ./analysis_poles > analysis_poles_cpp.csv

#include "ctrlpp/model/analysis.h"
#include "ctrlpp/model/discretize.h"
#include "ctrlpp/model/state_space.h"

#include <algorithm>
#include <cstdio>

int main()
{
    using Scalar = double;
    constexpr std::size_t NX = 2;
    constexpr std::size_t NU = 1;
    constexpr std::size_t NY = 1;

    ctrlpp::continuous_state_space<Scalar, NX, NU, NY> sys_c{};
    sys_c.A << 0.0, 1.0, -1.0, -0.5;
    sys_c.B << 0.0, 1.0;
    sys_c.C << 1.0, 0.0;
    sys_c.D << 0.0;

    auto sys_d = ctrlpp::discretize(ctrlpp::zoh{}, sys_c, 0.05);

    auto pc = ctrlpp::poles(sys_c);
    auto pd = ctrlpp::poles(sys_d);

    // Sort by imaginary part descending
    std::sort(pc.begin(), pc.end(), [](auto a, auto b) { return a.imag() > b.imag(); });
    std::sort(pd.begin(), pd.end(), [](auto a, auto b) { return a.imag() > b.imag(); });

    bool ctrb = ctrlpp::is_controllable<Scalar, NX, NU>(sys_d.A, sys_d.B);
    bool obsv = ctrlpp::is_observable<Scalar, NX, NY>(sys_d.A, sys_d.C);
    bool stable_c = ctrlpp::is_stable(sys_c);
    bool stable_d = ctrlpp::is_stable(sys_d);

    std::printf("pc0_re,pc0_im,pc1_re,pc1_im,pd0_re,pd0_im,pd1_re,pd1_im,ctrb_rank,obsv_rank,stable_c,stable_d\n");
    std::printf("%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e,%.15e\n",
                pc[0].real(), pc[0].imag(), pc[1].real(), pc[1].imag(),
                pd[0].real(), pd[0].imag(), pd[1].real(), pd[1].imag(),
                ctrb ? 2.0 : 1.0, obsv ? 2.0 : 1.0,
                stable_c ? 1.0 : 0.0, stable_d ? 1.0 : 0.0);
}
